Add field to allow skipping of tests

ycombinator opened the pull request:
This PR adds a field to the pipeline and system test configuration files, allowing tests to be skipped.

Related: https://github.com/elastic/elastic-package/issues/218

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-spec%2FPR-105/detail/PR-105/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-spec%2FPR-105/detail/PR-105/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-spec%2FPR-105/detail/PR-105/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-spec%2FPR-105/detail/PR-105/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://package-spec_105.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `Pull request #105 opened`

* Start Time: 2021-01-12T20:09:11.427+0000

* Duration: 2 min 0 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->





### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-spec%2FPR-105/detail/PR-105/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>

  
  
  
  
  
##### `Check`
<ul>
<li>Took 0 min 5 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/package-spec/pipelines/PR-105/runs/1/steps/122/log/?start=0">here</a></li>
<li>Description: <code>make check</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-01-12T20:10:39.741Z] Wiping out workspace first.
[2021-01-12T20:10:39.750Z] Cloning the remote Git repository
[2021-01-12T20:10:39.750Z] Using shallow clone with depth 5
[2021-01-12T20:10:39.750Z] Avoid fetching tags
[2021-01-12T20:10:40.569Z] Cleaning workspace
[2021-01-12T20:10:40.582Z] Using shallow fetch with depth 5
[2021-01-12T20:10:40.582Z] Pruning obsolete local branches
[2021-01-12T20:10:39.766Z] Cloning repository git@github.com:elastic/package-spec.git
[2021-01-12T20:10:39.789Z]  > git init /var/lib/jenkins/workspace/gest-manager_package-spec_PR-105/src/github.com/elastic/package-spec # timeout=10
[2021-01-12T20:10:39.795Z] Fetching upstream changes from git@github.com:elastic/package-spec.git
[2021-01-12T20:10:39.795Z]  > git --version # timeout=10
[2021-01-12T20:10:39.803Z]  > git --version # 'git version 2.17.1'
[2021-01-12T20:10:39.803Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-01-12T20:10:39.808Z]  > git fetch --no-tags --progress -- git@github.com:elastic/package-spec.git +refs/heads/*:refs/remotes/origin/* # timeout=15
[2021-01-12T20:10:40.551Z]  > git config remote.origin.url git@github.com:elastic/package-spec.git # timeout=10
[2021-01-12T20:10:40.554Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2021-01-12T20:10:40.563Z]  > git config remote.origin.url git@github.com:elastic/package-spec.git # timeout=10
[2021-01-12T20:10:40.571Z]  > git rev-parse --verify HEAD # timeout=10
[2021-01-12T20:10:40.574Z] No valid HEAD. Skipping the resetting
[2021-01-12T20:10:40.574Z]  > git clean -fdx # timeout=10
[2021-01-12T20:10:40.584Z] Fetching upstream changes from git@github.com:elastic/package-spec.git
[2021-01-12T20:10:40.584Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-01-12T20:10:40.589Z]  > git fetch --no-tags --progress --prune -- git@github.com:elastic/package-spec.git +refs/pull/105/head:refs/remotes/origin/PR-105 +refs/heads/master:refs/remotes/origin/master # timeout=15
[2021-01-12T20:10:41.149Z] Merging remotes/origin/master commit 930b44726a527ac599f55eebdfca114f049bdeff into PR head commit 57305f90c105a20eaf64a07c23bcafe43548e88d
[2021-01-12T20:10:41.218Z] Merge succeeded, producing 57305f90c105a20eaf64a07c23bcafe43548e88d
[2021-01-12T20:10:41.218Z] Checking out Revision 57305f90c105a20eaf64a07c23bcafe43548e88d (PR-105)
[2021-01-12T20:10:41.151Z]  > git config core.sparsecheckout # timeout=10
[2021-01-12T20:10:41.154Z]  > git checkout -f 57305f90c105a20eaf64a07c23bcafe43548e88d # timeout=15
[2021-01-12T20:10:41.192Z]  > git remote # timeout=10
[2021-01-12T20:10:41.196Z]  > git config --get remote.origin.url # timeout=10
[2021-01-12T20:10:41.200Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-01-12T20:10:41.204Z]  > git merge 930b44726a527ac599f55eebdfca114f049bdeff # timeout=10
[2021-01-12T20:10:41.211Z]  > git rev-parse HEAD^{commit} # timeout=10
[2021-01-12T20:10:41.220Z]  > git config core.sparsecheckout # timeout=10
[2021-01-12T20:10:41.223Z]  > git checkout -f 57305f90c105a20eaf64a07c23bcafe43548e88d # timeout=15
[2021-01-12T20:10:41.262Z] Commit message: "Add field to allow skipping of tests"
[2021-01-12T20:10:41.262Z] Cleaning workspace
[2021-01-12T20:10:41.316Z]  > git --version # timeout=10
[2021-01-12T20:10:41.319Z]  > git --version # 'git version 2.17.1'
[2021-01-12T20:10:41.323Z] fatal: bad object fadb851f608d4b87127fd41bbb51427835afc4a3
[2021-01-12T20:10:41.264Z]  > git rev-parse --verify HEAD # timeout=10
[2021-01-12T20:10:41.267Z] Resetting working tree
[2021-01-12T20:10:41.267Z]  > git reset --hard # timeout=10
[2021-01-12T20:10:41.290Z]  > git clean -fdx # timeout=10
[2021-01-12T20:10:42.301Z] Masking supported pattern matches of $GIT_USERNAME or $GIT_PASSWORD
[2021-01-12T20:10:43.040Z] + git fetch https://****:****@github.com/elastic/package-spec.git +refs/pull/*/head:refs/remotes/origin/pr/*
[2021-01-12T20:10:43.690Z] Running in /var/lib/jenkins/workspace/gest-manager_package-spec_PR-105/src/github.com/elastic/package-spec/.git
[2021-01-12T20:10:43.723Z] Archiving artifacts
[2021-01-12T20:10:44.402Z] + git rev-parse HEAD
[2021-01-12T20:10:44.734Z] + git rev-parse HEAD
[2021-01-12T20:10:45.057Z] + git rev-parse origin/pr/105
[2021-01-12T20:10:45.107Z] [INFO] githubEnv: Found Git Build Cause: pr
[2021-01-12T20:10:45.455Z] Masking supported pattern matches of $GITHUB_TOKEN
[2021-01-12T20:10:46.320Z] [WARN] githubApiCall: The REST API call https://api.github.com/repos/elastic/package-spec/pulls/105/reviews return 0 elements
[2021-01-12T20:10:46.372Z] [INFO] githubPrCheckApproved: Title: Add field to allow skipping of tests - User: ycombinator - Author Association: MEMBER
[2021-01-12T20:10:47.313Z] Stashed 297 file(s)
[2021-01-12T20:10:47.400Z] Running in /var/lib/jenkins/workspace/gest-manager_package-spec_PR-105/src/github.com/elastic/package-spec
[2021-01-12T20:10:48.442Z] + gvm 1.14.2
[2021-01-12T20:10:55.030Z] export GOROOT="/var/lib/jenkins/workspace/gest-manager_package-spec_PR-105/.gvm/versions/go1.14.2.linux.amd64"
[2021-01-12T20:10:55.030Z] export PATH="/var/lib/jenkins/workspace/gest-manager_package-spec_PR-105/.gvm/versions/go1.14.2.linux.amd64/bin:$PATH"
[2021-01-12T20:10:55.404Z] + go get -u github.com/magefile/mage
[2021-01-12T20:10:57.693Z] + go get -u github.com/elastic/go-licenser
[2021-01-12T20:10:59.012Z] + go get -u golang.org/x/tools/cmd/goimports
[2021-01-12T20:11:04.675Z] + go get -u github.com/jstemmer/go-junit-report
[2021-01-12T20:11:05.713Z] Running in /var/lib/jenkins/workspace/gest-manager_package-spec_PR-105/src/github.com/elastic/package-spec
[2021-01-12T20:11:06.033Z] + make check
[2021-01-12T20:11:06.033Z] make[1]: Entering directory '/var/lib/jenkins/workspace/gest-manager_package-spec_PR-105/src/github.com/elastic/package-spec/code/go'
[2021-01-12T20:11:06.033Z] go: downloading golang.org/x/lint v0.0.0-20201208152925-83fdc39ff7b5
[2021-01-12T20:11:06.033Z] go: found golang.org/x/lint/golint in golang.org/x/lint v0.0.0-20201208152925-83fdc39ff7b5
[2021-01-12T20:11:06.033Z] go: downloading golang.org/x/tools v0.0.0-20200130002326-2f3ba24bd6e7
[2021-01-12T20:11:06.605Z] go: golang.org/x/tools upgrade => v0.0.0-20210112183307-1e6ecd4bf1b0
[2021-01-12T20:11:06.868Z] go: downloading golang.org/x/tools v0.0.0-20210112183307-1e6ecd4bf1b0
[2021-01-12T20:11:08.787Z] go: downloading github.com/rakyll/statik v0.1.7
[2021-01-12T20:11:08.787Z] go: downloading github.com/xeipuuv/gojsonreference v0.0.0-20180127040603-bd5ef7bd5415
[2021-01-12T20:11:08.787Z] go: downloading github.com/pkg/errors v0.9.1
[2021-01-12T20:11:08.787Z] go: downloading gopkg.in/yaml.v3 v3.0.0-20200313102051-9f266ea9e77c
[2021-01-12T20:11:08.787Z] go: downloading github.com/creasty/defaults v1.5.1
[2021-01-12T20:11:08.787Z] go: downloading github.com/xeipuuv/gojsonschema v1.2.0
[2021-01-12T20:11:08.787Z] go: downloading github.com/Masterminds/semver/v3 v3.1.0
[2021-01-12T20:11:08.787Z] go: downloading github.com/xeipuuv/gojsonpointer v0.0.0-20180127040702-4e3ac2762d5f
[2021-01-12T20:11:09.053Z] go: downloading github.com/stretchr/testify v1.6.1
[2021-01-12T20:11:09.053Z] go: downloading gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405
[2021-01-12T20:11:09.317Z] go: downloading github.com/davecgh/go-spew v1.1.0
[2021-01-12T20:11:09.317Z] go: downloading github.com/pmezard/go-difflib v1.0.0
[2021-01-12T20:11:10.267Z] Specs are not up to date. Please run make update.
[2021-01-12T20:11:10.267Z] Makefile:21: recipe for target 'check-spec' failed
[2021-01-12T20:11:10.267Z] make[1]: *** [check-spec] Error 1
[2021-01-12T20:11:10.267Z] make[1]: Leaving directory '/var/lib/jenkins/workspace/gest-manager_package-spec_PR-105/src/github.com/elastic/package-spec/code/go'
[2021-01-12T20:11:10.267Z] Makefile:7: recipe for target 'check' failed
[2021-01-12T20:11:10.267Z] make: *** [check] Error 2
[2021-01-12T20:11:10.413Z] Stage "Test" skipped due to earlier failure(s)
[2021-01-12T20:11:10.703Z] Running on Jenkins in /var/lib/jenkins/workspace/gest-manager_package-spec_PR-105
[2021-01-12T20:11:10.849Z] [INFO] getVaultSecret: Getting secrets
[2021-01-12T20:11:10.951Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-01-12T20:11:11.716Z] + chmod 755 generate-build-data.sh
[2021-01-12T20:11:11.716Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/package-spec/PR-105/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/package-spec/PR-105/runs/1 FAILURE 120017
[2021-01-12T20:11:11.966Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/package-spec/PR-105/runs/1/steps/?limit=10000 -o steps-info.json
[2021-01-12T20:11:12.517Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/package-spec/PR-105/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-01-12T20:11:12.767Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2021-01-12T20:11:13.679Z] Retry 2/3 exited 22, retrying in 2 seconds...

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

mtojek reviewed (commented):
I'm not in favor of skipping tests, but we may need it in the future. I would enforce the description and link with regexp :) Something like these (all fields required):

```yaml
skip:
   url: 
   explanation:
```

WDYT?

ycombinator commented:
@mtojek thanks, I like this idea. It forces users to offer a short explanation for the skip and URL that has more details. 👍

ycombinator commented:
@mtojek I addressed your review feedback. A few points to note:
- I went with `link` instead of `url` because we already have a precedent for `link` here: https://github.com/elastic/package-spec/blob/930b44726a527ac599f55eebdfca114f049bdeff/versions/1/changelog.spec.yml#L37
- I went with `reason` instead of `explanation` because `reason` is shorter. 😄 
- I don't like that the definition of `skip` is repeated in two places. I tried refactoring it into a common file, similar to what we have with https://github.com/elastic/package-spec/blob/930b44726a527ac599f55eebdfca114f049bdeff/versions/1/data_stream/manifest.spec.yml#L35 but I was not successful. I kept getting a `Reference /spec.yml#/definitions/skip must be canonical`.
  - The value of my `$ref` property looked like `"../spec.yml/#definitions/skip"` and in the `data_stream/_dev/test/spec.yml` file I had a `spec.definitions.skip` property.
  - Just for testing I even tried to create a `common.yml` file in the same folder as the referencing spec file and changed the value of `$ref` to `"./common.yml/#definitions/skip"` but that also gave the same error.
  - If you have some ideas as to what might be causing this error or what else I should try, please let me know.

mtojek commented:
I looked into this problem and found a workaround: https://github.com/mtojek/package-spec/tree/test-skip-fix-1

Most likely the inline content doesn't allow for jumping over remote resources or the location is not evaluated correctly.

ycombinator commented:
Thanks @mtojek. I merged your workaround into this PR. Now I'm seeing a different test failure, which I'm investigating:

```
=== RUN   TestValidateFile/good
    validator_test.go:50:
                Error Trace:    validator_test.go:50
                Error:          Received unexpected error:
                                found 1 validation error:
                                   1. item [test-access-event.json-config.json] is not allowed in folder [../../internal/validator/test/packages/good/data_stream/foo/_dev/test/pipeline]
                Test:           TestValidateFile/good
```
```

ycombinator commented:
@mtojek This PR is ready for your re-review when you get a chance. Thanks!

mtojek commented on versions/1/data_stream/_dev/test/pipeline/config_raw.spec.yml:
```diff
@@ -0,0 +1,20 @@
+spec:
```
I think the config is applicable in both modes now: raw and JSON. It's described also in docs for pipeline tests.

ycombinator replied:
There appears to be a slight difference in the configs for the raw and JSON modes.

Raw: https://github.com/elastic/package-spec/blob/930b44726a527ac599f55eebdfca114f049bdeff/versions/1/data_stream/_dev/test/pipeline/spec.yml#L14-L36

JSON: https://github.com/elastic/package-spec/blob/930b44726a527ac599f55eebdfca114f049bdeff/versions/1/data_stream/_dev/test/pipeline/spec.yml#L37-L55

It looks like the raw mode supports `multiline` whereas the JSON mode does not. 

That's why I had to end up creating two different spec files: `config_raw.spec.yml` and `config_json.spec.yml`.

mtojek replied:
Ah, I see! In this case I think it's correct. Sorry about confusion.
